abi/dynamic/reflect_marshal.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/consts"
+)
+
+func DynamicMarshal(inputAbi abi.ABI, typeName string, jsonData string) ([]byte, error) {
+	// Find the type in the ABI
+	abiType := findABIType(inputAbi, typeName)
+	if abiType == nil {
+		return nil, fmt.Errorf("type %s not found in ABI", typeName)
+	}
+
+	// Create a cache to avoid rebuilding types
+	typeCache := make(map[string]reflect.Type)
+
+	// Create a dynamic struct type
+	dynamicType := getReflectType(typeName, inputAbi, typeCache)
+
+	// Create an instance of the dynamic struct
+	dynamicValue := reflect.New(dynamicType).Interface()
+
+	// Unmarshal JSON data into the dynamic struct
+	if err := json.Unmarshal([]byte(jsonData), dynamicValue); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal JSON data: %w", err)
+	}
+
+	// Marshal the dynamic struct using LinearCodec
+	writer := codec.NewWriter(0, consts.NetworkSizeLimit)
+	if err := codec.LinearCodec.MarshalInto(dynamicValue, writer.Packer); err != nil {
+		return nil, fmt.Errorf("failed to marshal struct: %w", err)
+	}
+
+	return writer.Bytes(), nil
+}
+
+func DynamicUnmarshal(inputAbi abi.ABI, typeName string, data []byte) (string, error) {
+	// Find the type in the ABI
+	abiType := findABIType(inputAbi, typeName)
+	if abiType == nil {
+		return "", fmt.Errorf("type %s not found in ABI", typeName)
+	}
+
+	// Create a cache to avoid rebuilding types
+	typeCache := make(map[string]reflect.Type)
+
+	// Create a dynamic struct type
+	dynamicType := getReflectType(typeName, inputAbi, typeCache)
+
+	// Create an instance of the dynamic struct
+	dynamicValue := reflect.New(dynamicType).Interface()
+
+	// Unmarshal the data into the dynamic struct
+	if err := codec.LinearCodec.Unmarshal(data, dynamicValue); err != nil {
+		return "", fmt.Errorf("failed to unmarshal data: %w", err)
+	}
+
+	// Marshal the dynamic struct back to JSON
+	jsonData, err := json.Marshal(dynamicValue)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal struct to JSON: %w", err)
+	}
+
+	return string(jsonData), nil
+}
+
+func getReflectType(abiTypeName string, inputAbi abi.ABI, typeCache map[string]reflect.Type) reflect.Type {
+	switch abiTypeName {
+	case "string":
+		return reflect.TypeOf("")
+	case "uint8":
+		return reflect.TypeOf(uint8(0))
+	case "uint16":
+		return reflect.TypeOf(uint16(0))
+	case "uint32":
+		return reflect.TypeOf(uint32(0))
+	case "uint64":
+		return reflect.TypeOf(uint64(0))
+	case "int8":
+		return reflect.TypeOf(int8(0))
+	case "int16":
+		return reflect.TypeOf(int16(0))
+	case "int32":
+		return reflect.TypeOf(int32(0))
+	case "int64":
+		return reflect.TypeOf(int64(0))
+	case "Address":
+		return reflect.TypeOf(codec.Address{})
+	default:
+		if strings.HasPrefix(abiTypeName, "[]") {
+			elemType := getReflectType(strings.TrimPrefix(abiTypeName, "[]"), inputAbi, typeCache)
+			return reflect.SliceOf(elemType)
+		} else if strings.HasPrefix(abiTypeName, "[") {
```

Comment:
(nit, subjective style) The `else` is redundant because of the `return` on the line before. Later with the cache check you don't have one.

I think there should at least be consistency. Personally I'd remove it because it adds just that tiny bit of cognitive load for the reader who thinks "else what?".